---
title: "Detecting differential enrichment of H3K27me3 in the mouse lung epithelium"
author:
- name: Aaron T. L. Lun
  affiliation: 
  - &WEHI The Walter and Eliza Hall Institute of Medical Research, 1G Royal Parade, Parkville, VIC 3052, Melbourne, Australia
  - Department of Medical Biology, The University of Melbourne, Parkville, VIC 3010, Melbourne, Australia
- name: Aliaksei Holik
  affiliation:
  - *WEHI
date: "2019-06-04"
vignette: >
  %\VignetteIndexEntry{4. Differential enrichment of H3K27me3 in lung epithelium}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::html_document:
    toc_float: yes
    titlecaps: false
bibliography: ref.bib
---



# Overview

Here, we perform a window-based DB analysis to identify regions of differential H3K27me3 enrichment in mouse lung epithelium.
H3K27me3 is associated with transcriptional repression and is usually observed with broad regions of enrichment.
The aim of this workflow is to demonstrate how to analyze these broad marks with *[csaw](https://bioconductor.org/packages/3.10/csaw)*,
especially at variable resolutions with multiple window sizes.
We use H3K27me3 ChIP-seq data from a study comparing wild-type (WT) and _Ezh2_ knock-out (KO) animals [@galvis2015repression],
contains two biological replicates for each genotype.
We download BAM files and indices using *[chipseqDBData](https://bioconductor.org/packages/3.10/chipseqDBData)*.


```r
library(chipseqDBData)
h3k27me3data <- H3K27me3Data()
h3k27me3data
```

```
## DataFrame with 4 rows and 3 columns
##          Name                 Description
##   <character>                 <character>
## 1  SRR1274188        control H3K27me3 (1)
## 2  SRR1274189        control H3K27me3 (2)
## 3  SRR1274190 Ezh2 knock-out H3K27me3 (1)
## 4  SRR1274191 Ezh2 knock-out H3K27me3 (2)
##                                              Path
##                                       <character>
## 1 /tmp/RtmpxiwDuH/file60f6717fa646/SRR1274188.bam
## 2 /tmp/RtmpxiwDuH/file60f6717fa646/SRR1274189.bam
## 3 /tmp/RtmpxiwDuH/file60f6717fa646/SRR1274190.bam
## 4 /tmp/RtmpxiwDuH/file60f6717fa646/SRR1274191.bam
```

# Pre-processing checks

We check some mapping statistics with *[Rsamtools](https://bioconductor.org/packages/3.10/Rsamtools)*.


```r
library(Rsamtools)
diagnostics <- list()
for (bam in h3k27me3data$Path) {
    total <- countBam(bam)$records
    mapped <- countBam(bam, param=ScanBamParam(
        flag=scanBamFlag(isUnmapped=FALSE)))$records
    marked <- countBam(bam, param=ScanBamParam(
        flag=scanBamFlag(isUnmapped=FALSE, isDuplicate=TRUE)))$records
    diagnostics[[basename(bam)]] <- c(Total=total, Mapped=mapped, Marked=marked)
}
diag.stats <- data.frame(do.call(rbind, diagnostics))
diag.stats$Prop.mapped <- diag.stats$Mapped/diag.stats$Total*100
diag.stats$Prop.marked <- diag.stats$Marked/diag.stats$Mapped*100
diag.stats
```

```
##                   Total   Mapped  Marked Prop.mapped Prop.marked
## SRR1274188.bam 24445704 18605240 2769679    76.10842    14.88655
## SRR1274189.bam 21978677 17014171 2069203    77.41217    12.16164
## SRR1274190.bam 26910067 18606352 4361393    69.14272    23.44034
## SRR1274191.bam 21354963 14092438 4392541    65.99140    31.16949
```

We construct a `readParam` object to standardize the parameter settings in this analysis.
For consistency with the original analysis by @galvis2015repression,
we will define the blacklist using the predicted repeats from the RepeatMasker software.


```r
library(BiocFileCache)
bfc <- BiocFileCache("local", ask=FALSE)
black.path <- bfcrpath(bfc, file.path("http://hgdownload.cse.ucsc.edu",
    "goldenPath/mm10/bigZips/chromOut.tar.gz"))
tmpdir <- tempfile()
dir.create(tmpdir)
untar(black.path, exdir=tmpdir)

# Iterate through all chromosomes.
collected <- list()
for (x in list.files(tmpdir, full=TRUE)) {
    f <- list.files(x, full=TRUE, pattern=".fa.out")
    to.get <- vector("list", 15)
    to.get[[5]] <- "character"
    to.get[6:7] <- "integer"
    collected[[length(collected)+1]] <- read.table(f, skip=3, 
        stringsAsFactors=FALSE, colClasses=to.get)
}

collected <- do.call(rbind, collected)
blacklist <- GRanges(collected[,1], IRanges(collected[,2], collected[,3]))
blacklist
```

```
## GRanges object with 5147737 ranges and 0 metadata columns:
##                         seqnames          ranges strand
##                            <Rle>       <IRanges>  <Rle>
##         [1]                 chr1 3000001-3002128      *
##         [2]                 chr1 3003153-3003994      *
##         [3]                 chr1 3003994-3004054      *
##         [4]                 chr1 3004041-3004206      *
##         [5]                 chr1 3004207-3004270      *
##         ...                  ...             ...    ...
##   [5147733] chrY_JH584303_random   152557-155890      *
##   [5147734] chrY_JH584303_random   155891-156883      *
##   [5147735] chrY_JH584303_random   157070-157145      *
##   [5147736] chrY_JH584303_random   157909-157960      *
##   [5147737] chrY_JH584303_random   157953-158099      *
##   -------
##   seqinfo: 66 sequences from an unspecified genome; no seqlengths
```

We set the minimum mapping quality score to 10 to remove poorly or non-uniquely aligned reads.


```r
library(csaw)
param <- readParam(minq=10, discard=blacklist)
param
```

```
##     Extracting reads in single-end mode
##     Duplicate removal is turned off 
##     Minimum allowed mapping score is 10 
##     Reads are extracted from both strands
##     No restrictions are placed on read extraction
##     Reads in 5147737 regions will be discarded
```

# Counting reads into windows

Reads are then counted into sliding windows using *[csaw](https://bioconductor.org/packages/3.10/csaw)* [@lun2015csaw].
At this stage, we use a large 2 kbp window to reflect the fact that H3K27me3 exhibits broad enrichment.
This allows us to increase the size of the counts and thus detection power,
without having to be concerned about loss of genomic resolution to detect sharp binding events.


```r
win.data <- windowCounts(h3k27me3data$Path, param=param, width=2000,
    spacing=500, ext=200)
win.data
```

```
## class: RangedSummarizedExperiment 
## dim: 4614717 4 
## metadata(6): spacing width ... param final.ext
## assays(1): counts
## rownames: NULL
## rowData names(0):
## colnames: NULL
## colData names(4): bam.files totals ext rlen
```

We use `spacing=500` to avoid redundant work when sliding a large window across the genome.
The default spacing of 50 bp would result in many windows with over 90% overlap in their positions,
increasing the amount of computational work without a meaningful improvement in resolution.
We also set the fragment length to 200 bp based on experimental knowledge of the size selection procedure.
Unlike the previous analyses, the fragment length cannot easily estimated here due to weak strand bimodality of diffuse marks.

# Normalization for composition biases

As in the CBP example, we normalize for composition biases resulting from imbalanced DB between conditions [@lun2014denovo].
We expect systematic DB in one direction as Ezh2 function (and thus some H3K27me3 deposition activity) is lost in the KO genotype.
We apply the TMM method [@robinson2010scaling] to counts for large 10 kbp bins,
and store the resulting normalization factors back in `win.data` for use in the DB analysis with the window counts.


```r
bins <- windowCounts(h3k27me3data$Path, bin=TRUE, width=10000, param=param)
win.data <- normFactors(bins, se.out=win.data)
(normfacs <- win.data$norm.factors)
```

```
## [1] 0.9946875 0.9984751 1.0080895 0.9987965
```

Figure \@ref(fig:compoplot) shows the effect of normalization on the relative enrichment between pairs of samples. 
We see that log-ratio of normalization factors passes through the centre of the cloud of background regions in each plot,
indicating that the bias has been successfully identified and removed.


```r
bin.ab <- scaledAverage(bins)
adjc <- calculateCPM(bins, use.norm.factors=FALSE)

par(cex.lab=1.5, mfrow=c(1,3))
smoothScatter(bin.ab, adjc[,1]-adjc[,2], ylim=c(-6, 6),
    xlab="Average abundance", ylab="Log-ratio (1 vs 2)")
abline(h=log2(normfacs[1]/normfacs[4]), col="red")

smoothScatter(bin.ab, adjc[,1]-adjc[,3], ylim=c(-6, 6),
    xlab="Average abundance", ylab="Log-ratio (1 vs 3)")
abline(h=log2(normfacs[2]/normfacs[4]), col="red")

smoothScatter(bin.ab, adjc[,1]-adjc[,4], ylim=c(-6, 6),
    xlab="Average abundance", ylab="Log-ratio (1 vs 4)")
abline(h=log2(normfacs[3]/normfacs[4]), col="red")
```

<div class="figure">
<img src="h3k27me3_files/figure-html/compoplot-1.png" alt="Mean-difference plots for the bin counts, comparing sample 1 to all other samples. The red line represents the log-ratio of the normalization factors between samples." width="1152" />
<p class="caption">(\#fig:compoplot)Mean-difference plots for the bin counts, comparing sample 1 to all other samples. The red line represents the log-ratio of the normalization factors between samples.</p>
</div>

# Filtering of low-abundance windows

We estimate the global background and remove low-abundance windows that are not enriched above this background level.
To retain a window, we require it to have at least 2-fold more coverage than the average background.
This is less stringent than the thresholds used in previous analyses, owing the weaker enrichment observed for diffuse marks.


```r
filter.stat <- filterWindows(win.data, bins, type="global")
min.fc <- 2 
```

Figure \@ref(fig:bghistplot) shows that chosen threshold is greater than the abundances of most bins in the genome,
presumably those corresponding to background regions.
This suggests that the filter will remove most windows lying within background regions.


```r
hist(filter.stat$back.abundances, main="", breaks=50,
    xlab="Background abundance (log2-CPM)")
threshold <- filter.stat$abundances[1] - filter.stat$filter[1] + log2(min.fc)
abline(v=threshold, col="red")
```

<div class="figure">
<img src="h3k27me3_files/figure-html/bghistplot-1.png" alt="Histogram of average abundances across all 10 kbp genomic bins. The filter threshold is shown as the red line." width="100%" />
<p class="caption">(\#fig:bghistplot)Histogram of average abundances across all 10 kbp genomic bins. The filter threshold is shown as the red line.</p>
</div>

The majority of windows in background regions are filtered out upon applying a modest fold-change threshold.
This leaves a small set of relevant windows for further analysis.


```r
keep <- filter.stat$filter > log2(min.fc)
summary(keep)
```

```
##    Mode   FALSE    TRUE 
## logical 4555355   59362
```

```r
filtered.data <- win.data[keep,]
```

# Statistical modelling of biological variability

Counts for each window are modelled using *[edgeR](https://bioconductor.org/packages/3.10/edgeR)* [@mccarthy2012differential; @robinson2010edger].
We first convert our `RangedSummarizedExperiment` object into a `DGEList`.


```r
library(edgeR)
y <- asDGEList(filtered.data)
str(y)
```

```
## Formal class 'DGEList' [package "edgeR"] with 1 slot
##   ..@ .Data:List of 2
##   .. ..$ : int [1:59362, 1:4] 17 19 32 33 30 31 31 36 33 29 ...
##   .. .. ..- attr(*, "dimnames")=List of 2
##   .. .. .. ..$ : chr [1:59362] "1" "2" "3" "4" ...
##   .. .. .. ..$ : chr [1:4] "Sample1" "Sample2" "Sample3" "Sample4"
##   .. ..$ :'data.frame':	4 obs. of  3 variables:
##   .. .. ..$ group       : Factor w/ 1 level "1": 1 1 1 1
##   .. .. ..$ lib.size    : int [1:4] 10602308 9971590 8605634 5659274
##   .. .. ..$ norm.factors: num [1:4] 0.995 0.998 1.008 0.999
```

We then construct a design matrix for our experimental design.
Here, we use a simple one-way layout with two groups of two replicates.


```r
genotype <- h3k27me3data$Description
genotype[grep("control", genotype)] <- "wt"
genotype[grep("knock-out", genotype)] <- "ko"

genotype <- factor(genotype)
design <- model.matrix(~0+genotype)
colnames(design) <- levels(genotype)
design
```

```
##   ko wt
## 1  0  1
## 2  0  1
## 3  1  0
## 4  1  0
## attr(,"assign")
## [1] 1 1
## attr(,"contrasts")
## attr(,"contrasts")$genotype
## [1] "contr.treatment"
```

We estimate the negative binomial (NB) and quasi-likelihood (QL) dispersions for each window [@lund2012ql].
We again observe an increasing trend in the NB dispersions with respect to abundance (Figure \@ref(fig:bcvplot)).


```r
y <- estimateDisp(y, design)
summary(y$trended.dispersion)
```

```
##     Min.  1st Qu.   Median     Mean  3rd Qu.     Max. 
## 0.008979 0.009484 0.010091 0.010970 0.010446 0.025661
```

```r
plotBCV(y)
```

<div class="figure">
<img src="h3k27me3_files/figure-html/bcvplot-1.png" alt="Abundance-dependent trend in the BCV for each window, represented by the blue line. Common (red) and tagwise estimates (black) are also shown." width="100%" />
<p class="caption">(\#fig:bcvplot)Abundance-dependent trend in the BCV for each window, represented by the blue line. Common (red) and tagwise estimates (black) are also shown.</p>
</div>

The QL dispersions are strongly shrunk towards the trend (Figure \@ref(fig:qlplot)), 
indicating that there is little variability in the dispersions across windows.


```r
fit <- glmQLFit(y, design, robust=TRUE)
summary(fit$df.prior)
```

```
##    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
##  0.4548 54.7389 54.7389 54.7155 54.7389 54.7389
```

```r
plotQLDisp(fit)
```

<div class="figure">
<img src="h3k27me3_files/figure-html/qlplot-1.png" alt="Effect of EB shrinkage on the raw QL dispersion estimate for each window (black) towards the abundance-dependent trend (blue) to obtain squeezed estimates (red)." width="100%" />
<p class="caption">(\#fig:qlplot)Effect of EB shrinkage on the raw QL dispersion estimate for each window (black) towards the abundance-dependent trend (blue) to obtain squeezed estimates (red).</p>
</div>

These results are consistent with the presence of systematic differences in enrichment between replicates,
as discussed in the [CBP analysis](https://bioconductor.org/packages/3.10/chipseqDB/vignettes/cbp.html#statistical-modelling-of-biological-variability).
Nonetheless, the samples separate by genotype in the MDS plot (Figure \@ref(fig:mdsplot)),
which suggests that the downstream analysis will be able to detect DB regions.


```r
plotMDS(cpm(y, log=TRUE), top=10000, labels=genotype,
    col=c("red", "blue")[as.integer(genotype)])
```

<div class="figure">
<img src="h3k27me3_files/figure-html/mdsplot-1.png" alt="MDS plot with two dimensions for all samples in the H3K27me3 data set. Samples are labelled and coloured according to the genotype. A larger top set of windows was used to improve the visualization of the genome-wide differences between the WT samples." width="100%" />
<p class="caption">(\#fig:mdsplot)MDS plot with two dimensions for all samples in the H3K27me3 data set. Samples are labelled and coloured according to the genotype. A larger top set of windows was used to improve the visualization of the genome-wide differences between the WT samples.</p>
</div>

We then test for DB between conditions in each window using the QL F-test.


```r
contrast <- makeContrasts(wt-ko, levels=design)
res <- glmQLFTest(fit, contrast=contrast)
```

# Consolidating results from multiple window sizes

Consolidation allows the analyst to incorporate information from a range of different window sizes,
each of which has a different trade-off between resolution and count size.
This is particularly useful for broad marks where the width of an enriched region can be variable,
as can the width of the differentially bound interval of an enriched region.
To demonstrate, we repeat the entire analysis using 500 bp windows.
Compared to our previous 2 kbp analysis, this provides greater spatial resolution at the cost of lowering the counts.


```r
# Counting into 500 bp windows.
win.data2 <- windowCounts(h3k27me3data$Path, param=param, width=500,
    spacing=100, ext=200)

# Re-using the same normalization factors.
win.data2$norm.factors <- win.data$norm.factors

# Filtering on abundance.
filter.stat2 <- filterWindows(win.data2, bins, type="global")
keep2 <- filter.stat2$filter > log2(min.fc)
filtered.data2 <- win.data2[keep2,]

# Performing the statistical analysis.
y2 <- asDGEList(filtered.data2)
y2 <- estimateDisp(y2, design)
fit2 <- glmQLFit(y2, design, robust=TRUE)
res2 <- glmQLFTest(fit2, contrast=contrast)
```

We consolidate the 500 bp analysis with our previous 2 kbp analysis using the `mergeWindowsList()` function.
This clusters both sets of windows together into a single set of regions.
To limit chaining effects, each region cannot be more than 30 kbp in size.


```r
merged <- mergeResultsList(list(filtered.data, filtered.data2), 
    tab.list=list(res$table, res2$table),
    equiweight=TRUE, tol=100, merge.args=list(max.width=30000))
merged$regions
```

```
## GRanges object with 80389 ranges and 0 metadata columns:
##           seqnames            ranges strand
##              <Rle>         <IRanges>  <Rle>
##       [1]     chr1   3050001-3052500      *
##       [2]     chr1   3086401-3087200      *
##       [3]     chr1   3098001-3098800      *
##       [4]     chr1   3139501-3140100      *
##       [5]     chr1   3211501-3212600      *
##       ...      ...               ...    ...
##   [80385]     chrY 90766001-90769500      *
##   [80386]     chrY 90774801-90775900      *
##   [80387]     chrY 90788501-90789000      *
##   [80388]     chrY 90793101-90793700      *
##   [80389]     chrY 90797001-90814000      *
##   -------
##   seqinfo: 66 sequences from an unspecified genome
```

We compute combined $p$-values using Simes' method for region-level FDR control [@simes1986; @lun2014denovo].
This is done after weighting the contributions from the two sets of windows
to ensure that the combined $p$-value for each region is not dominated by the analysis with more (smaller) windows.


```r
tabcom <- merged$combined
is.sig <- tabcom$FDR <= 0.05
summary(is.sig)
```

```
##    Mode   FALSE    TRUE 
## logical   73687    6914
```

Ezh2 is one of the proteins responsible for depositing H3K27me3, 
so we might expect that most DB regions have increased enrichment in the WT condition.
However, the opposite seems to be true here, which is an interesting result that may warrant further investigation. 


```r
table(tabcom$direction[is.sig])
```

```
## 
## down   up 
## 4599 2315
```

We also obtain statistics for the window with the lowest $p$-value in each region.
The signs of the log-fold changes are largely consistent with the `direction` numbers above.


```r
tabbest <- merged$best
is.sig.pos <- (tabbest$logFC > 0)[is.sig]
summary(is.sig.pos)
```

```
##    Mode   FALSE    TRUE 
## logical    4599    2315
```

Finally, we save these results to file for future reference.


```r
out.ranges <- merged$regions
mcols(out.ranges) <- data.frame(tabcom,
    best.logFC=tabbest$logFC)
saveRDS(file="h3k27me3_results.rds", out.ranges)
```

# Annotation and visualization

We add annotation for each region using the `detailRanges()` function, 
as [previously described](https://bioconductor.org/packages/3.10/chipseqDB/vignettes/h3k9ac.html#using-the-detailranges-function).


```r
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
anno <- detailRanges(out.ranges, orgdb=org.Mm.eg.db,
    txdb=TxDb.Mmusculus.UCSC.mm10.knownGene)
mcols(out.ranges) <- cbind(mcols(out.ranges), anno)
```

We visualize one of the DB regions overlapping the _Cdx2_ gene to reproduce the results in @holik2015transcriptome.


```r
cdx2 <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)["12591"] # Cdx2 Entrez ID
cur.region <- subsetByOverlaps(out.ranges, cdx2)[1]
cur.region
```

```
## GRanges object with 1 range and 10 metadata columns:
##       seqnames              ranges strand |  nWindows  logFC.up logFC.down
##          <Rle>           <IRanges>  <Rle> | <integer> <integer>  <integer>
##   [1]     chr5 147299501-147308000      * |        41        41          0
##                     PValue                  FDR   direction       best.logFC
##                  <numeric>            <numeric> <character>        <numeric>
##   [1] 1.60546113388456e-06 0.000333509723845953          up 2.96437327774961
##           overlap        left       right
##       <character> <character> <character>
##   [1]   Cdx2:-:PE                        
##   -------
##   seqinfo: 66 sequences from an unspecified genome
```



We use *[Gviz](https://bioconductor.org/packages/3.10/Gviz)* [@hahne2016visualizing] to plot the results.
As in the H3K9ac analysis, we set up some tracks to display genome coordinates and gene annotation.


```r
library(Gviz)
gax <- GenomeAxisTrack(col="black", fontsize=15, size=2)
greg <- GeneRegionTrack(TxDb.Mmusculus.UCSC.mm10.knownGene, showId=TRUE,
    geneSymbol=TRUE, name="", background.title="transparent")
symbols <- unlist(mapIds(org.Mm.eg.db, gene(greg), "SYMBOL",
    "ENTREZID", multiVals = "first"))
symbol(greg) <- symbols[gene(greg)]
```

In Figure \@ref(fig:tfplot), we see enrichment of H3K27me3 in the WT condition at the _Cdx2_ locus.
This is consistent with the known regulatory relationship between Ezh2 and _Cdx2_.


```r
collected <- list()
lib.sizes <- filtered.data$totals/1e6
for (i in seq_along(h3k27me3data$Path)) {
    reads <- extractReads(bam.file=h3k27me3data$Path[i], cur.region, param=param)
    cov <- as(coverage(reads)/lib.sizes[i], "GRanges")
    collected[[i]] <- DataTrack(cov, type="histogram", lwd=0, ylim=c(0,1),
        name=h3k27me3data$Description[i], col.axis="black", col.title="black",
        fill="darkgray", col.histogram=NA)
}

plotTracks(c(gax, collected, greg), chromosome=as.character(seqnames(cur.region)),
    from=start(cur.region), to=end(cur.region))
```

<div class="figure">
<img src="h3k27me3_files/figure-html/tfplot-1.png" alt="Coverage tracks for a region with H3K27me3 enrichment in KO (top two tracks) against the WT (last two tracks)." width="768" />
<p class="caption">(\#fig:tfplot)Coverage tracks for a region with H3K27me3 enrichment in KO (top two tracks) against the WT (last two tracks).</p>
</div>

In contrast, if we look at a constitutively expressed gene, such as _Col1a2_, we find little evidence of H3K27me3 deposition.
Only a single window is retained after filtering on abundance,
and this window shows no evidence of changes in H3K27me3 enrichment between WT and KO samples.


```r
col1a2 <- genes(TxDb.Mmusculus.UCSC.mm10.knownGene)["12843"] # Col1a2 Entrez ID
cur.region <- subsetByOverlaps(out.ranges, col1a2)
cur.region
```

```
## GRanges object with 1 range and 10 metadata columns:
##       seqnames          ranges strand |  nWindows  logFC.up logFC.down
##          <Rle>       <IRanges>  <Rle> | <integer> <integer>  <integer>
##   [1]     chr6 4511701-4512200      * |         1         0          1
##                 PValue               FDR   direction         best.logFC
##              <numeric>         <numeric> <character>          <numeric>
##   [1] 0.22213538480528 0.435745190943813        down -0.712940757651797
##           overlap         left        right
##       <character>  <character>  <character>
##   [1]  Col1a2:+:I Col1a2:+:907 Col1a2:+:161
##   -------
##   seqinfo: 66 sequences from an unspecified genome
```



# Session information


```r
sessionInfo()
```

```
## R version 3.6.0 Patched (2019-05-02 r76458)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 18.04.2 LTS
## 
## Matrix products: default
## BLAS:   /home/luna/Software/R/R-3-6-branch-dev/lib/libRblas.so
## LAPACK: /home/luna/Software/R/R-3-6-branch-dev/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] grid      stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] Gviz_1.29.0                             
##  [2] org.Mm.eg.db_3.8.2                      
##  [3] TxDb.Mmusculus.UCSC.mm10.knownGene_3.4.7
##  [4] GenomicFeatures_1.37.1                  
##  [5] AnnotationDbi_1.47.0                    
##  [6] edgeR_3.27.4                            
##  [7] limma_3.41.2                            
##  [8] csaw_1.19.2                             
##  [9] SummarizedExperiment_1.15.1             
## [10] DelayedArray_0.11.0                     
## [11] BiocParallel_1.19.0                     
## [12] matrixStats_0.54.0                      
## [13] Biobase_2.45.0                          
## [14] BiocFileCache_1.9.0                     
## [15] dbplyr_1.4.0                            
## [16] Rsamtools_2.1.2                         
## [17] Biostrings_2.53.0                       
## [18] XVector_0.25.0                          
## [19] GenomicRanges_1.37.8                    
## [20] GenomeInfoDb_1.21.1                     
## [21] IRanges_2.19.6                          
## [22] S4Vectors_0.23.4                        
## [23] BiocGenerics_0.31.2                     
## [24] chipseqDBData_1.1.0                     
## [25] knitr_1.23                              
## [26] BiocStyle_2.13.0                        
## 
## loaded via a namespace (and not attached):
##  [1] colorspace_1.4-1              biovizBase_1.33.0            
##  [3] htmlTable_1.13.1              base64enc_0.1-3              
##  [5] dichromat_2.0-0               rstudioapi_0.10              
##  [7] bit64_0.9-7                   interactiveDisplayBase_1.23.0
##  [9] codetools_0.2-16              splines_3.6.0                
## [11] Formula_1.2-3                 cluster_2.0.9                
## [13] shiny_1.3.2                   BiocManager_1.30.4           
## [15] compiler_3.6.0                httr_1.4.0                   
## [17] backports_1.1.4               assertthat_0.2.1             
## [19] Matrix_1.2-17                 lazyeval_0.2.2               
## [21] later_0.8.0                   acepack_1.4.1                
## [23] htmltools_0.3.6               prettyunits_1.0.2            
## [25] tools_3.6.0                   gtable_0.3.0                 
## [27] glue_1.3.1                    GenomeInfoDbData_1.2.1       
## [29] dplyr_0.8.1                   rappdirs_0.3.1               
## [31] Rcpp_1.0.1                    ExperimentHub_1.11.1         
## [33] rtracklayer_1.45.1            xfun_0.7                     
## [35] stringr_1.4.0                 mime_0.6                     
## [37] ensembldb_2.9.1               statmod_1.4.32               
## [39] XML_3.98-1.19                 AnnotationHub_2.17.3         
## [41] zlibbioc_1.31.0               scales_1.0.0                 
## [43] BSgenome_1.53.0               VariantAnnotation_1.31.3     
## [45] ProtGenerics_1.17.2           hms_0.4.2                    
## [47] promises_1.0.1                AnnotationFilter_1.9.0       
## [49] RColorBrewer_1.1-2            yaml_2.2.0                   
## [51] curl_3.3                      memoise_1.1.0                
## [53] gridExtra_2.3                 ggplot2_3.1.1                
## [55] biomaRt_2.41.0                rpart_4.1-15                 
## [57] latticeExtra_0.6-28           stringi_1.4.3                
## [59] RSQLite_2.1.1                 highr_0.8                    
## [61] checkmate_1.9.3               rlang_0.3.4                  
## [63] pkgconfig_2.0.2               bitops_1.0-6                 
## [65] evaluate_0.14                 lattice_0.20-38              
## [67] purrr_0.3.2                   GenomicAlignments_1.21.2     
## [69] htmlwidgets_1.3               bit_1.1-14                   
## [71] tidyselect_0.2.5              plyr_1.8.4                   
## [73] magrittr_1.5                  bookdown_0.11                
## [75] R6_2.4.0                      Hmisc_4.2-0                  
## [77] DBI_1.0.0                     pillar_1.4.1                 
## [79] foreign_0.8-71                survival_2.44-1.1            
## [81] RCurl_1.95-4.12               nnet_7.3-12                  
## [83] tibble_2.1.2                  crayon_1.3.4                 
## [85] KernSmooth_2.23-15            rmarkdown_1.13               
## [87] progress_1.2.2                locfit_1.5-9.1               
## [89] data.table_1.12.2             blob_1.1.1                   
## [91] digest_0.6.19                 xtable_1.8-4                 
## [93] httpuv_1.5.1                  munsell_0.5.0
```

# References

